### PR TITLE
feat(nsis): support choosing compression algorithms, closes #7685

### LINKED ISSUE #7685: [feat] tauri conf to set NSIS bundler options - SetCompressor
### Describe the problem

My app contains a pyinstaller onefolder applicaiton as resource ,When I change build target from 'msi' to 'nsis', Build artifacts have become smaller，from 112MB to 70MB , but the installer decompression   slowly on client machine!

### Describe the solution you'd like

provide a option in tauri.conf.json as "nsis" chidren ,allow developer to set compressor to three kind;
https://nsis.sourceforge.io/Reference/SetCompressor
e.g:
{
  "tauri":{
      "bundle":{
          "windows":{
              "nsis":{"setCompressor":"zlib|bzip2|lzma"}              
}    
}
}
}

SetCompressor /SOLID zlib|bzip2|lzma

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by pewsheen**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [x] Yes, and the changes were approved in issue #7685 
- [ ] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### CODE REVIEW

File: core/tauri-utils/src/config.rs [hunk 1]
```diff
@@ -480,6 +493,8 @@ pub struct NsisConfig {
   /// By default the OS language is selected, with a fallback to the first language in the `languages` array.
   #[serde(default, alias = "display-language-selector")]
   pub display_language_selector: bool,
+  /// Set the compression algorithm used to compress files in the installer.
+  pub set_compressor: Option<NsisCompressor>,
```
Comment (amrbashir) [inline core/tauri-utils/src/config.rs]:
```suggestion
  pub compression: Option<NsisCompressor>,
```

File: tooling/bundler/src/bundle/settings.rs [hunk 2]
```diff
@@ -289,6 +289,8 @@ pub struct NsisSettings {
   /// Whether to display a language selector dialog before the installer and uninstaller windows are rendered or not.
   /// By default the OS language is selected, with a fallback to the first language in the `languages` array.
   pub display_language_selector: bool,
+  /// Set compression algorithm used in the installer.
+  pub set_compressor: Option<NsisCompressor>,
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/settings.rs]:
```suggestion
  /// Set compression algorithm used to compress files in the installer.
  pub compression: Option<NsisCompressor>,
```

File: core/tauri-utils/src/config.rs [hunk 3]
```diff
@@ -438,6 +438,19 @@ pub struct WixConfig {
   pub dialog_image_path: Option<PathBuf>,
 }
 
+/// Compression algorithms used in the NSIS installer.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
+#[cfg_attr(feature = "schema", derive(JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub enum NsisCompressor {
```
Comment (amrbashir) [inline core/tauri-utils/src/config.rs]:
```suggestion
pub enum NsisCompression {
```

File: core/tauri-utils/src/config.rs [hunk 4]
```diff
@@ -438,6 +438,19 @@ pub struct WixConfig {
...
+  /// ZLIB compressor
+  Zlib,
+  /// BZIP2 compressor
+  Bzip2,
+  /// LZMA compressor (default)
```
Comment (amrbashir) [inline core/tauri-utils/src/config.rs]:
```suggestion
  /// ZLIB uses the deflate algorithm, it is a quick and simple method. With the default compression level it uses about 300 KB of memory.
  Zlib,
  /// BZIP2 usually gives better compression ratios than ZLIB, but it is a bit slower and uses more memory. With the default compression level it uses about 4 MB of memory.
  Bzip2,
  /// LZMA (default) is anew compression method that gives very good compression ratios. The decompression speed is high (10-20 MB/s on a 2 GHz CPU), the compression speed is lower. The memory size that will be used for decompression is the dictionary size plus a few KBs, the default is 8 MB.
```

File: core/tauri-utils/src/config.rs [hunk 5]
```diff
@@ -438,6 +438,19 @@ pub struct WixConfig {
   pub dialog_image_path: Option<PathBuf>,
 }
 
+/// Compression algorithms used in the NSIS installer.
```
Comment (amrbashir) [inline core/tauri-utils/src/config.rs:441]:
```suggestion
/// Compression algorithms used in the NSIS installer.
///
/// See <https://nsis.sourceforge.io/Reference/SetCompressor>
```

File: core/tauri-utils/src/config.rs [hunk 6]
```diff
@@ -480,6 +493,8 @@ pub struct NsisConfig {
   /// By default the OS language is selected, with a fallback to the first language in the `languages` array.
   #[serde(default, alias = "display-language-selector")]
   pub display_language_selector: bool,
+  /// Set the compression algorithm used to compress files in the installer.
```
Comment (amrbashir) [inline core/tauri-utils/src/config.rs:498]:
```suggestion
  /// Set the compression algorithm used to compress files in the installer.
  ///
  /// See <https://nsis.sourceforge.io/Reference/SetCompressor>
```

File: tooling/bundler/src/bundle/windows/nsis.rs [hunk 7]
```diff
@@ -243,6 +243,16 @@ fn build_nsis_app_installer(
         to_json(dunce::canonicalize(sidebar_image)?),
       );
     }
+    if let Some(compressor) = &nsis.set_compressor {
+      data.insert(
+        "set_compressor",
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/nsis.rs]:
```suggestion
        "compression",
```

File: tooling/bundler/src/bundle/windows/templates/installer.nsi [hunk 8]
```diff
@@ -28,12 +28,19 @@
 !define WEBVIEW2INSTALLERPATH "{{webview2_installer_path}}"
 !define UNINSTKEY "Software\Microsoft\Windows\CurrentVersion\Uninstall\${PRODUCTNAME}"
 !define MANUPRODUCTKEY "Software\${MANUFACTURER}\${PRODUCTNAME}"
+!define SETCOMPRESSOR "{{set_compressor}}"
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/templates/installer.nsi]:
```suggestion
!define COMPRESSION "{{compression}}"
```

File: tooling/bundler/src/bundle/windows/nsis.rs [hunk 9]
```diff
@@ -243,6 +243,16 @@ fn build_nsis_app_installer(
...
+        to_json(match compressor {
+          NsisCompressor::Zlib => "zlib",
+          NsisCompressor::Bzip2 => "bzip2",
+          NsisCompressor::Lzma => "lzma",
+        }),
+      );
+    }
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/nsis.rs]: Just use the default here
```suggestion
    data.insert(
      "set_compressor",
      to_json(match compressor.unwrap_or(NsisCompressor::Lzma) {
        NsisCompressor::Zlib => "zlib",
        NsisCompressor::Bzip2 => "bzip2",
        NsisCompressor::Lzma => "lzma",
      }),
    );
    
```

File: tooling/bundler/src/bundle/windows/templates/installer.nsi [hunk 10]
```diff
@@ -28,12 +28,19 @@
...
 OutFile "${OUTFILE}"
 Unicode true
-SetCompressor /SOLID lzma
+
+; Set the compression algorithm. Default is LZMA.
+!if "${SETCOMPRESSOR}" == ""
+  SetCompressor /SOLID lzma
+!else
+  SetCompressor /SOLID "${SETCOMPRESSOR}"
+!endif
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/templates/installer.nsi]:
```suggestion
; Set the compression algorithm. Default is LZMA.
SetCompressor /SOLID "${SETCOMPRESSOR}"
```

**Review by amrbashir — CHANGES_REQUESTED** [review]
Just a couple of docs changes and naming consistency

File: tooling/bundler/src/bundle/windows/nsis.rs (hunk 9 above)
Comment (pewsheen) [reply tooling/bundler/src/bundle/windows/nsis.rs]: nice

File: tooling/bundler/src/bundle/windows/nsis.rs (hunk 7 above)
Comment (pewsheen) [reply tooling/bundler/src/bundle/windows/nsis.rs]: fixed

File: tooling/bundler/src/bundle/windows/templates/installer.nsi [hunk 11]
```diff
@@ -28,12 +28,19 @@
...
 OutFile "${OUTFILE}"
 Unicode true
-SetCompressor /SOLID lzma
+
+; Set the compression algorithm. Default is LZMA.
+!if "${COMPRESSION}" == ""
+  SetCompressor /SOLID lzma
+!else
+  SetCompressor /SOLID "${COMPRESSION}"
+!endif
```
Comment (pewsheen) [inline tooling/bundler/src/bundle/windows/templates/installer.nsi:43]: I just found out that we still need this condition.

If the user doesn't provide `nsis` configuration in `tauri.config.json`, the bundler will not run the builder logic in `nsis.rs`, which means the `{{compression}}` won't be assigned a default value (`NsisCompression::Lzma`) hence `"${COMPRESSION}" == ""`

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:43]: Right! Let's leave it like that then

**Review by amrbashir — CHANGES_REQUESTED** [review]
Just noticed that this was opened against `dev`, let's rebase and open it against `1.x` so we could roll it out in 1.5

**Comment by pewsheen** [comment]
> Just noticed that this was opened against `dev`, let's rebase and open it against `1.x` so we could roll it out in 1.5

omg. It's a mess now

I'll close this one and create another PR later.

**Pull request closed without merging**